include/coap/str.h

```diff
@@ -39,6 +44,32 @@ coap_string_t *coap_new_string(size_t size);
  */
 void coap_delete_string(coap_string_t *string);
 
+/**
+ * Returns a new const string object with at least size+1 bytes storage
+ * allocated, and the provided data copied into the string object.
+ * The string must be released using coap_delete_string_const();
+ *
+ * @param size The size to allocate for the binary string data
+ *
+ * @return       A pointer to the new object or @c NULL on error.
+ */
+coap_str_const_t *coap_new_str_const(const uint8_t *data, size_t size);
+
+/**
+ * Deletes the given const string and releases any memory allocated.
+ *
+ * @param string The string to free off
+ */
+void coap_delete_str_const(coap_str_const_t *string);
+
+/**
+ * Take the specified byte array (text) and create a coap_str_const_t *
+ *
+ * @param string The const byte array to convert to a coap_str_const_t *
+ */
+#define coap_make_str_const(string) \
+  (&(coap_str_const_t){.length = sizeof(string)-1,.s = (const uint8_t *)(string)})
```

Comment:
Here, we have a pitfall:

```c
void foo(const char *foo) {
  coap_make_str_const(foo);
  ...
}
```
`.length` now will be 7 (or 3, respectively).

Reply:
Hmm.  Good catch.  Can make it strlen(string) instead of sizeof(string)-1, but that will not handle binary text.  I'll look to see if there is a way to do this without having to count the characters.

Reply:
This would not work for binary data anyway as the length would be wrong. How about using `strlen()` and making explicit that this is for zero-terminated character strings only?
(And possibly have a second variant for binary data plus size.)

Reply:
Certainly can explicitly state that this is for zero terminated char strings.  A binary alternative using sizeof() can be created, but it will have the same limitations of your foo() example.

Perhaps we just need to state that results from the usage will be "random" if not used for an array of bytes defined in the local scope.